test/HcalDigiPipelineTest.cxx

```diff
@@ -0,0 +1,286 @@
+
+    const auto recHits = event.getCollection<ldmx::HcalHit>("HcalRecHits");
+    CHECK(recHits.size() == 1);
+
+    auto hit = recHits.at(0);
+    ldmx::HcalID id(hit.getID());
+    CHECK_FALSE(hit.isNoise());
+    CHECK(id.raw() == simHits.at(0).getID());
+
+    ntuple_.setVar<float>("RecX", hit.getXPos());
+    ntuple_.setVar<float>("RecY", hit.getYPos());
+    ntuple_.setVar<float>("RecZ", hit.getZPos());
+    
+    // define target energy by using the settings at the top
+    auto target_daq_energy = Approx(truth_energy).epsilon(MAX_ENERGY_PERCENT_ERROR_DAQ_TOT_MODE/100);
+    if (is_in_adc_mode) target_daq_energy = Approx(truth_energy).margin(MAX_ENERGY_ERROR_DAQ_ADC_MODE);
+
+    CHECK(hit.getEnergy() == target_daq_energy);
+    ntuple_.setVar<float>("RecEnergy", hit.getEnergy());
+        
+    return;
+  }
+}; // HcalCheckEnergyReconstruction
+
+/**
+ * @class HcalCheckPositionReconstruction
+ * Checks:
+ * - Position of HcalRecHit matches SimCalorimeterHit with the same ID 
+ * The SimHits are all generated at the same energy (1 MIP) for consistency.
+ */
+class HcalCheckPositionReconstruction : public framework::Analyzer {
+
+public:
+  HcalCheckPositionReconstruction(const std::string &name, framework::Process &p)
+      : framework::Analyzer(name, p) {}
+  ~HcalCheckPositionReconstruction() {}
+
+  void onProcessStart() final override {
+  }
+
+  void analyze(const framework::Event &event) final override {
+    const auto simHits = event.getCollection<ldmx::SimCalorimeterHit>("HcalSimHits");
+
+    CHECK(simHits.size() > 0);
+
+    for(int ihit=0; ihit<simHits.size(); ihit++) {
+      auto hit = simHits.at(ihit);
+      double x = hit.getPosition()[0];
+      double y = hit.getPosition()[1];
+      double z = hit.getPosition()[2];
+    }
+
+    return;
+  }
+}; // HcalCheckPositionReconstruction 
+  
+} // namespace test
+} // namespace hcal
+
+DECLARE_PRODUCER_NS(hcal::test, HcalFakeSimHits)
+DECLARE_ANALYZER_NS(hcal::test, HcalCheckEnergyReconstruction)
+DECLARE_ANALYZER_NS(hcal::test, HcalCheckPositionReconstruction)
+
+/**
+ * Test for the Hcal Digi Pipeline
+ *
+ * Does not check for realism. Simply makes sure sim energies
+ * end up being "close" to output rec energies.
+ *
+ * Checks
+ *  - Keep reconstructed energy deposited close to simulated value (with ADC readout mode)
+ *  - Keep reconstructed positions close to simulated or expected value
+ *
+ * @TODO still need to expand to multiple contribs in a single sim hit
+ * @TODO check with TOT mode when implemented
+ */
+TEST_CASE("Hcal Digi Pipeline test", "[Hcal][functionality]") {
+
+  const std::string config_file{"hcal_digi_pipeline_test_config.py"};
+
+  char **args;
+  framework::ProcessHandle p;
+
+  framework::ConfigurePython cfg(config_file, args, 0);
+  REQUIRE_NOTHROW(p = cfg.makeProcess());
+  p->run();
+}
+TEST_CASE("Hcal Geometry test", "[Hcal][functionality]") {
+
+  const std::string config_file{"hcal_geometry_test_config.py"};
```

Comment:
Where is this file? It doesn't look like it is in your commit tree...

Reply:
Right, so I was testing this offline. I added this [config file](https://github.com/LDMX-Software/Hcal/blob/5a7474b035dfff9b716b795a30c07e008987efc3/test/hcal_geometry_test_config.py) and the [analyzer](https://github.com/LDMX-Software/Hcal/blob/5a7474b035dfff9b716b795a30c07e008987efc3/test/HcalGeometryTest.cxx). The config generates muons for a rectangle or rather a cube, that I think covers the volume of the whole Hcal starting from the Ecal_z position so that it has coverage over the sideHcal. This basic test does not really test for all the reconstructed positions but for the ones returned by the HcalGeometry class, and checks that they are compatible within 20% (you can vary this number but point you cannot be too precise in some cases).

Reply:
For now though I have [commented it out](https://github.com/LDMX-Software/Hcal/blob/5a7474b035dfff9b716b795a30c07e008987efc3/test/HcalGeometryTest.cxx#L111) from the test sequence because when I try to run over both the DigiPipeline and the Geometry test I get this error (after running the Digi test correctly): 
`-------------------------------------------------------------------------------
Hcal Geometry test
-------------------------------------------------------------------------------
/Users/cmantill/LDMX/Jan31/ldmx-sw/Hcal/test/HcalGeometryTest.cxx:100
...............................................................................

/Users/cmantill/LDMX/Jan31/ldmx-sw/Hcal/test/HcalGeometryTest.cxx:107: FAILED:
  {Unknown expression after the reported line}
due to a fatal error condition:
  SIGSEGV - Segmentation violation signal

===============================================================================
test cases:    2 |    0 passed |   2 failed
assertions: 2032 | 1931 passed | 101 failed`
(I had to change the FakeSimHits name too so that they don't interfere but I still get an error).

@tomeichlersmith  do you have any suggestions to integrate both tests? Otherwise we can leave the Geometry test out.

Reply:
So, I've seen this error for a wide variety of bugs. Some things to check:
- Make sure python script doesn't have an error in it. (Try `ldmx python3 config.py` to make sure it can load it)
- Make sure you can run each test individually (With `run_test` you can [specify the tests to run](https://github.com/catchorg/Catch2/blob/devel/docs/command-line.md#specifying-which-tests-to-run). Do this to make sure that it is failing because they are being run together.

Reply:
Hi @tomeichlersmith, I tried both of your suggestions. `ldmx python3 config` works fine for both configs and when specifying both tests separately they run successfully (some failed cases in the digi pipeline because I increased the # of PEs when testing). Do you have other suggestions to debug?

```bash-3.2$ 
ldmx ../run_test [Hcal][functionality]
Filters: [Hcal][functionality]
===============================================================================
test cases:    1 |    0 passed |   1 failed
assertions: 6001 | 5737 passed | 264 failed
ldmx ../run_test [Hcal][Geometry]
Filters: [Hcal][Geometry]
===============================================================================
All tests passed (1430 assertions in 1 test case)
```

Reply:
@tomeichlersmith it seems that what could be causing this to fail is the ntupleProducer (active only in the hcal_digi_pipeline_test). For now when I comment that (in the latest [commit](https://github.com/LDMX-Software/Hcal/pull/5/commits/0371a035d332740128f758bd95aecacbb9b0ff3c)) both tests run with no problem.